datafusion/optimizer/src/optimizer.rs

```diff
@@ -223,6 +223,12 @@ impl Optimizer {
                 let result = rule.try_optimize(&new_plan, optimizer_config);
                 match result {
                     Ok(Some(plan)) => {
+                        if plan.schema() != new_plan.schema() {
+                            return Err(DataFusionError::Internal(format!(
```

Comment:
Could we include the expected and actual schema in the error message? This will make debugging much easier.

Reply:
Also, could you add a unit test?